common/raw_hashtable_metadata_group.h

```diff
@@ -0,0 +1,930 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef CARBON_COMMON_RAW_HASHTABLE_METADATA_GROUP_H_
+#define CARBON_COMMON_RAW_HASHTABLE_METADATA_GROUP_H_
+
+#include <cstddef>
+#include <cstring>
+#include <iterator>
+
+#include "common/check.h"
+#include "llvm/ADT/Sequence.h"
+#include "llvm/ADT/bit.h"
+#include "llvm/Support/FormatVariadic.h"
+#include "llvm/Support/MathExtras.h"
+
+// Detect whether we can use SIMD accelerated implementations of the control
+// groups.
+#if defined(__SSSE3__)
+#include <x86intrin.h>
+#define CARBON_X86_SIMD_SUPPORT 1
+#elif defined(__ARM_NEON)
+#include <arm_neon.h>
+#define CARBON_NEON_SIMD_SUPPORT 1
+#endif
+
+// A namespace collecting a set of low-level utilities for building hashtable
+// data structures. These should only be used as implementation details of
+// higher-level data structure APIs.
+//
+// The utilities in this file are:
+//
+// - Primitives to manage "groups" of hashtable entries that have densely packed
+//   control bytes we can scan rapidly as a group, often using SIMD facilities
+//   to process the entire group at once.
+namespace Carbon::RawHashtable {
+
+// We define a constant max group size. The particular group size used in
+// practice may vary, but we want to have some upper bound that can be reliably
+// used when allocating memory to ensure we don't create fractional groups and
+// memory allocation is done consistently across the architectures.
+constexpr ssize_t MaxGroupSize = 16;
+
+// An index encoded as low zero bits ending in (at least) one set high bit. The
+// index can be extracted by counting the low zero bits. It's presence can be
+// tested directly however by checking for any zero bits. The underlying type to
+// be used is provided as `MaskT` which must be an unsigned integer type.
```

Comment:
I think moving away from "Mask" terminology for this type and these values would be clearer. I think these are much more like "match (bit) sets."

Reply:
Switched to "bits" here, and "match_bits" below.